Arrangement_on_surface_2/doc/Arrangement_on_surface_2/CGAL/Arr_caching_polyline_traits_2.h

```diff
@@ -0,0 +1,166 @@
+ * \cgalModels `ArrangementDirectionalXMonotoneTraits_2`
+ * \cgalModels `ArrangementConstructXMonotoneCurveTraits_2`
+ * \cgalModels `ArrangementApproximateTraits_2`
+ *
+ * \sa `Arr_polyline_traits_2<SubcurveTraits_2>`
+ */
+
+template <typename Kernel, typename Range>
+class Arr_caching_polyline_traits_2 :
+    public Arr_polycurve_traits_2<SegmentTraits_2> {
+public:
+
+  /// \name Types
+  /// @{
+  /*!
+   */
+  typedef typename Kernel::Point_2                                Point_2;
+
+  /*!
+   */
+  typedef internal::X_monotone_caching_polyline_2<Kernel, Range>
+    X_monotone_curve_2;
+
+  /*!
+   */
+  typedef internal::X_monotone_caching_polyline_2<Kernel, Range>  Curve_2;
+  typedef
+  /// @}
+
+  /*! Construction functor of a polyline. Its `operator()` is oveloaded to
+   * support various input types.
+   */
+  class Construct_curve_2 {
+  public:
+    /// \name Operations
+    /// @{
+
+    /*! Obtain a polyline connecting the two given endpoints.
+     * \param p the first point.
+     * \param q the second point.
+     * \pre `p` and `q` are distinct.
+     * \return a polyline that consists of a single segment connecting `p` and
+     *         `q`.
+     */
+    Curve_2 operator()(const Point_2& p, const Point_2& q) const;
+
+    /*! Construct a well-oriented polyline from a range of points.
+     *
+     * \param range a range of points
+     * \param duplicate_first indicates whether the first point should be
+     *        duplicated to construct a closed polyline.
+     * \pre contains no duplicated points.
+     * \return a polyline that consists of the given range of points.
+     */
+    template <typename Range>
+    Curve_2 operator()(const Range& range, bool duplicate_first = false) const;
+
+    /// @} /* end of operations */
+  }; /* end of Arr_caching_polyline_traits_2::Construct_curve_2 */
+
+  /*! Construction functor of a \f$x\f$-monotone polyline. Its `operator()` is
+   * oveloaded to support various input types.
+   */
+  class Construct_x_monotone_curve_2 {
+  };
+  public:
+    /// \name Operations
+    /// @{
+
+    /*! Obtain a polyline connecting the two given endpoints.
+     * \param p the first point.
+     * \param q the second point.
+     * \pre `p` and `q` are distinct.
+     * \return a polyline that consists of a single segment connecting `p` and
+     *         `q`.
+     */
+    X_monotone_curve_2 operator()(const Point_2& p, const Point_2& q) const;
+
+    /*! Construct a well-oriented polyline from a range of points.
+     *
+     * \param range a range of points.
+     * \param duplicate_first indicates whether the first point should be
+     *        duplicated to construct a closed polyline.
+     * \pre contains no duplicated points.
+     * \return a polyline that consists of the given range of points.
+     */
+    template <typename Range>
+    X_monotone_curve_2 operator()(const Range& range,
+                                  bool duplicate_first = false) const;
```

Comment:
`X_monotone_curve_2` can't be constructed from a range (which is not guaranteed to be X monotone).

Reply:
I see you just added this overload in the code. I'm not sure how this is expected to be used, but I guess if there is precondition on the input range being X-monotone, it can't hurt.